_drafts/2023-08-15-debugging-android-devices.md

```diff
@@ -0,0 +1,274 @@
+---
+title: Debugging Android Devices
+description:
+  What should I include here... <CHRIS COME BACK AND FIX THIS!!>
+author: chayes
+---
+
+All hardware devices in the field experience bugs and need debugging. Android devices in specific are exceptionally complex with several hundred gigabytes of source code, dozens of components, and a wide range of uses.
+
+<!-- excerpt start -->
+
+In this article we will explore the different facilities and tools available to debug Android-based devices and produce robust systems that can handle a wide range of applications from smart fridges, to payment terminals, and of course mobile phones.
+
+<!-- excerpt end -->
+
+{% include newsletter.html %}
+
+{% include toc.html %}
+
+## Logging
+
+The first place any debugging starts is usually looking through logs. Android has two sets of logs, Logcat and Kernel logs.
+
+### Logcat
+
+Logcat logs are pretty straightforward, most Android app devs know about logcat There are circular buffers for main, system, crash, radio, and event logs for binaries running within the Android runtime. This can be system services, applications, and native binaries. The CLI allows you to filter specific packages and control the verbosity on a per-package basis.
+
+### Kernel Logs
+
+/proc/kmesg can be read to get the kernel-level logs. This is where you will find kernel boot logs and driver logs. You can also use the dmesg binary on the system to get nice colored output and access to several other useful options.
+
+### LNAV
+
+[lnav (The Logfile Navigator)](https://lnav.org/) is one of the most useful utilities I've ever found for reading Android logs. It’s a generic, but feature-rich log navigator on the CLI which supports multi-log interpolation. This means you can combine both the kernel and logcat logs into a single log viewer and understand what’s happening at both layers of the system. When working at Square, I worked on a product that was two discrete Android devices talking to each other over USB. I wrote a script to add a device identifier to each set of logs after the time stamp and was able to combine all 4 sets of log files into one view, thus debugging the entire product as a whole.
```

Comment:
I might sell the lnav story a bit more by explaining the format of Android logs, saying they're spammy and hard to read in plaintext first, and then sell multi-log interpolation as another added benefit. Or by explaining that kernel and logcat logs aren't together by default and combining them let you diagnose a problem you were having.

> It’s a generic

Probably worth specifying what this means here instead of having people find out below (that generic means not Android logcat specific)

Reply:
> Or by explaining that kernel and logcat logs aren't together by default and combining them let you diagnose a problem you were having that was harder otherwise

Isn't that already understood by the previous two sections explaining where to find each of them.

> Probably worth specifying what this means here instead of having people find out below (that generic means not Android logcat specific)

fair enough, moved it further up before talking about it being generic.